cli/command_shallowrestore.go

```diff
@@ -0,0 +1,147 @@
+package cli
+
+import (
+	"context"
+	"path/filepath"
+
+	"github.com/alecthomas/kingpin"
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/kopia/kopia/fs/localfs"
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/snapshot/restore"
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+)
+
+const (
+	shallowRestoreCommandHelp = `Restore one level of a directory or file from a snapshot into the
+specified target path where the subsequent directory levels and files
+will be represented by placeholder files with the .kopiadir or
+.kopiafile suffix as appropriate. Placeholder files will snapshot as
+if they were contents of the repository from which they were restored.
+This permits modifying a restored file tree without round-tripping
+files that do not require alteration.
+
+By default, the target path will be created by the shallowrestore
+command if it does not exist.
+
+The source to be restored is specified in the form of a directory or
+file ID and optionally a sub-directory path or a shallow placeholder
+file.
+
+For example, the following source and target arguments will
+shallowrestore the contents of the 'kffbb7c28ea6c34d6cbe555d1cf80faa9'
+directory into a new, local directory named 'd1'
+
+'shallowrestore kffbb7c28ea6c34d6cbe555d1cf80faa9 d1'
+
+Similarly, the following command will shallowrestore the contents of a
+subdirectory 'subdir/subdir2' under
+'kffbb7c28ea6c34d6cbe555d1cf80faa9' into a new, local directory named
+'sd2'
+
+'shallowrestore kffbb7c28ea6c34d6cbe555d1cf80faa9/subdir1/subdir2 sd2'
```

Comment:
thinking out loud here: 

What if we make `shallowrestore` purely about "expanding" previously written DirEntry placeholders and add flags to trigger shallow restore to main `restore` command?

the session would be something like:

```
# initial shallow restore
kopia restore kffbb7c28ea6c34d6cbe555d1cf80faa9 d1 --shallow-restore-at-depth=3 --shallow-restore-files=*.zip`

# this creates d1/somefile and removes d1/somefile.kopia-object
kopia restore-entry d1/somefile.kopia-object`

# this creates d1/somedir/ and removes d1/somedir.kopia-object
kopia restore-entry d1/somedir.kopia-object`
```

We might be able to do the reverse - snapshot a directory/file and replace with a placeholder. Wouldn't that be cool?

Reply:
> snapshot a directory/file and replace with placeholder. Wouldn't that be cool?

Yes. My original notion for this PR (before I tried to reduce it to the smallest viable subset) included a feature like that. I'd prefer to add "collapse" in a subsequent PR.

It doesn't seem too hard to let `kopia restore` do full (the default) or shallow (via a max depth flag). (I think `--shallow-restore-at-depth` is too long to type though -- maybe it could have an abbreviation too?) In which case, the current `shallowrestore` command becomes:

```
kopia restore kffbb7c28ea6c34d6cbe555d1cf80faa9 d1 --shallow-restore-at-depth=0
```

And:

```
kopia restore-entry d1/dir-or-file.kopia-object --shallow-restore-at-depth=0
```

would expand `d1/dir-or-file.kopia-object`. The way that this is currently implemented,

```
cd d1
kopia restore-entry dir-or-file.kopia-object
```

would work equivalently. (I'm not convinced that `restore-entry` is the clearest name. `reify-kopia-object` seems more precise to me but not particularly ergonomic. I'll see about providing an abbreviation.)


Later, add this command:

```
# in d1
kopia collapse-entry dir-or-file
```

It would make sure that `dir-or-file` is in the repository and then replace `dir-or-file` with the appropriate `dir-or-file.kopia-object` if it is.

Reply:
Done in f48c329